feat: tray & app menu item "Check for Updates..."

bajtos opened the pull request:
Add "Check for Updates..." to the Tray menu. On macOS, add this item to the Application menu too. While the check is running, replace "Check for Updates" with a disabled item "Checking for Updates".

See https://github.com/filecoin-project/filecoin-station/issues/53

The implementation is heavily inspired by IPFS Desktop:
- [src/auto-updater/index.js](https://github.com/ipfs/ipfs-desktop/blob/b269cde5231915d2930eefba55b706012843a155/src/auto-updater/index.js)
- [src/tray.js](https://github.com/ipfs/ipfs-desktop/blob/b269cde5231915d2930eefba55b706012843a155/src/tray.js)

### Out of scope

- Reworking the OS-level notification
- Showing a banner at the top of the web UI when an update is ready

### Screenshots

A new macOS application menu item:

<img width="264" alt="Screenshot 2022-07-27 at 14 48 56" src="https://user-images.githubusercontent.com/1140553/181251929-f6809950-00d1-4eae-ac9c-88f0caa59c38.png">

While checking for updates:

<img width="262" alt="Screenshot 2022-07-27 at 14 49 06" src="https://user-images.githubusercontent.com/1140553/181252035-ac722fca-25c9-4cd3-a61a-92dee4985571.png">

A new Tray menu item:

<img width="233" alt="Screenshot 2022-07-27 at 14 50 35" src="https://user-images.githubusercontent.com/1140553/181252104-9549a12b-ae41-49ec-897e-eb8bb838553c.png">

While checking for updates:

<img width="218" alt="Screenshot 2022-07-27 at 14 50 41" src="https://user-images.githubusercontent.com/1140553/181252150-bed56031-1513-4f78-8011-237ac3815d93.png">

Various dialogs shown by the app:

<img width="757" alt="Screenshot 2022-07-27 at 14 49 33" src="https://user-images.githubusercontent.com/1140553/181252217-5d764748-b726-47de-9722-3dc543763b0d.png">

<img width="327" alt="Screenshot 2022-07-27 at 14 51 43" src="https://user-images.githubusercontent.com/1140553/181252280-0b80bbbf-9df2-4d45-a456-380b21906391.png">
<img width="328" alt="Screenshot 2022-07-27 at 14 51 56" src="https://user-images.githubusercontent.com/1140553/181252288-eddea2f7-c23e-41fe-a306-631847d6ee0c.png">
<img width="312" alt="Screenshot 2022-07-27 at 14 52 05" src="https://user-images.githubusercontent.com/1140553/181252292-71c8da42-b5ae-420a-a421-6c7f52e0c139.png">

The annoying notification we should remove later:

<img width="357" alt="Screenshot 2022-07-27 at 14 51 24" src="https://user-images.githubusercontent.com/1140553/181252474-866d3a98-64b1-4009-8a1a-cd0dd7ab425a.png">

bajtos commented on main/app-menu.js:
```diff
@@ -0,0 +1,56 @@
+const { ipcMainEvents } = require('./ipc')
+
+function setupAppMenu (/** @type {import('./typings').Context} */ ctx) {
+  if (process.platform !== 'darwin') return
+
+  // Add "Check for updates..." item to the Application menu on MacOS
+  const menu = Menu.getApplicationMenu()
+  if (!menu) return
+
+  menu.items[0].submenu?.insert(1, new MenuItem({
+    id: 'checkForUpdates',
+    label: 'Check For Updates...',
+    click: () => { ctx.manualCheckForUpdates() }
+  }))
+  menu.items[0].submenu?.insert(2, new MenuItem({
+    id: 'checkingForUpdates',
+    label: 'Checking For Updates',
+    enabled: false,
+    visible: false
+  }))
```
I am not sure where other Electron applications place "Check for Updates..." item on Windows & Linux, so I am adding this feature to macOS only to keep things simple.

If you think we should add this item to the application menu on other platforms too, then let's create follow-up GH issues, one per platform.

juliangruber replied:
I think we should add these items there too 👍 

https://github.com/filecoin-project/filecoin-station/issues/63
https://github.com/filecoin-project/filecoin-station/issues/64

bajtos replied:
Thanks for creating the GH issues, @juliangruber 🙇🏻 I'll mention them in a code comment.

bajtos commented on main/updater.js:
```diff
@@ -1,69 +1,158 @@
+function checkForUpdatesInBackground () {
+  ipcMain.emit(ipcMainEvents.UPDATE_CHECK_STARTED)
+
+  // TODO: replace this with autoUpdater.checkForUpdatesAndNotify()
+  autoUpdater.checkForUpdates()
```
This todo comment is coming from the original source code. I am not sure if it's still relevant. Do you remember @lidel why you would like to switch from `autoUpdater.checkForUpdates()` to `autoUpdater.checkForUpdatesAndNotify()`?

lidel replied:
I vaguely remember that the rationale was that it is less code to maintain, and that meant smaller surface for breaking autoupdates when upstream APIs change in subtle ways (happened in the past)

juliangruber commented on main/app-menu.js:
```diff
@@ -0,0 +1,56 @@
+ */
+function setupIpcEventListeners (appMenu) {
+  ipcMain.on(ipcMainEvents.UPDATE_CHECK_STARTED, () => {
+    getItemById('checkForUpdates').enabled = false
```
```suggestion
```

I don't see the purpose of this, afaic if an item is invisible it's also disabled

juliangruber replied:
Same question applies to the tray configuration

bajtos replied:
To be honest, I am not sure about this either. I have simply copied what I found in ipfs-desktop:

https://github.com/ipfs/ipfs-desktop/blob/b269cde5231915d2930eefba55b706012843a155/src/tray.js#L305-L307

```js
    menu.getMenuItemById('checkForUpdates').enabled = !isUpdating
    menu.getMenuItemById('checkForUpdates').visible = !isUpdating
    menu.getMenuItemById('checkingForUpdates').visible = isUpdating
```

juliangruber commented on main/app-menu.js:
```diff
@@ -0,0 +1,56 @@
+    const item = appMenu.getMenuItemById(id)
+    if (!item) throw new Error(`Unknown app menu item id: ${id}`)
+    return item
+  }
```
Could we also do without this function? We'd get an error in the consuming lines anyway.

juliangruber replied:
Same question applies to the tray configuration

bajtos replied:
This function serves two purposes:
- Make the TypeScript compiler happy
- Report a more helpful error message than _Uncaught TypeError: Cannot set properties of undefined (setting 'visible')`_

juliangruber replied:
Is this even run through ts, given the .js extension?

bajtos replied:
Yes, we check types in JS files, that's why there are JSDoc comments in most places.

https://github.com/filecoin-project/filecoin-station/blob/62b5de84631c0bd3b01aa70407f4f1ccdf6ca021/tsconfig.json#L7-L8

I found this setup useful so far. I have full auto-completion support in VS Code and the type checker can catch many programming errors.

juliangruber commented on main/app-menu.js:
```diff
@@ -0,0 +1,56 @@
+function setupAppMenu (/** @type {import('./typings').Context} */ ctx) {
+  if (process.platform !== 'darwin') return
+
+  // Add "Check for updates..." item to the Application menu on MacOS
```
Did you consider Windows for this PR as well? If we don't add it now then an issue will help keeping the implementations in sync in the future

bajtos replied:
We are thinking alike 😀

https://github.com/filecoin-project/filecoin-station/pull/62#discussion_r931039942

juliangruber commented on main/updater.js:
```diff
@@ -1,69 +1,158 @@
+  setup(ctx)
+
+  checkForUpdatesInBackground() // async check on startup
+  setInterval(checkForUpdatesInBackground, 43200000) // every 12 hours
```
```suggestion
  setInterval(checkForUpdatesInBackground, ms('12h'))
```

https://github.com/vercel/ms helps clean up some comments

bajtos replied:
Personally, I am cautious about adding new dependencies to the project. Is the maintenance cost of keeping `ms` up to date worth the benefit of avoiding a comment?

BTW my PR is not changing this part, I just moved existing code & the comment.

juliangruber replied:
To me this is a DX improvement that overweighs the maintenance aspect, especially once we use it in more places. In my experience of using `ms` it hasn't been a maintenance burden.

bajtos commented:
@juliangruber I addressed your feedback in [172bf3b](https://github.com/filecoin-project/filecoin-station/pull/62/commits/172bf3bd193196a1d203c7d20a475bc5b783a509).

- Added a code comment about Windows & Linux
- Reworked update check interval `ms('12h')` 

LGTY now?

juliangruber commented on main/app-menu.js:
```diff
@@ -0,0 +1,57 @@
+    getItemById('checkingForUpdates').visible = false
+  })
+
+  /**
```
```suggestion
  /**
   * Get item or fail with a useful error message.
```

bajtos replied:
Done in [af82612](https://github.com/filecoin-project/filecoin-station/pull/62/commits/af8261296e09e04876d64135d36588b2360d2e52)
